src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -211,17 +211,21 @@ public static GraphItem<RemoteResolveResult> CreateUnresolvedMatch(LibraryRange
                     // check for the exact library through local repositories
                     var localMatch = await FindLibraryByVersionAsync(library, framework, localProviders, cacheContext, logger, cancellationToken);
 
-                    if (localMatch != null)
+                    if (localMatch != null
+                        && localMatch.Library.Version.Equals(library.Version))
```

Comment:
Here and [here](https://github.com/NuGet/NuGet.Client/pull/2583/files#diff-ab60ca67cbc52f0cd86ebbd3df3441e0R224) `RemoteMatch.Library` and `LibraryIdentity.Version` are both nullable.  They're public properties with a public getter and setter.  This code assumes that both properties --- `Library` and `Version` --- are not `null`.  While this assumption may be currently valid by convention, I think this change should be more defensive when dereferencing these properties.

Reply:
I understand the concern, but both instances `localMatch` as well as `remoteMatch` are constructed with `libraryIdentity` with version in it if they are constructed. We can't have a match without the library in it. And there are already bunch of checks down below (like [this](https://github.com/NuGet/NuGet.Client/pull/2583/files#diff-ab60ca67cbc52f0cd86ebbd3df3441e0L243)) 

I don't mind being more defensive here but feels it will not make any difference.

Reply:
discussed offline with Damon, and he's good with the existing code.